### PR TITLE
Fix Sisimai::SMTP::Error.soft_or_hard failure on onhold and unknown

### PR DESCRIPTION
**Pull request by lunatyq**

### CODE REVIEW

File: lib/sisimai/smtp/error.rb [hunk 1]
```diff
@@ -89,11 +89,13 @@ def soft_or_hard(argv1 = '', argv2 = '')
             getchecked = Sisimai::SMTP::Error.is_permanent(argv2)
 
             if getchecked.nil?
-              # The value is not defined (returned undef)
               softorhard = ''
             else
-              # The value is 0 or 1
-              softorhard = getchecked == 1 ? 'hard' : 'soft'
+              if getchecked
```
Comment (azumakuniyuki) [inline lib/sisimai/smtp/error.rb]: __[rubocop]__
* Style/ConditionalAssignment: Use the return of the conditional for variable assignment and comparison.

File: lib/sisimai/smtp/error.rb [hunk 2]
```diff
@@ -89,11 +89,13 @@ def soft_or_hard(argv1 = '', argv2 = '')
...
+                softorhard = 'hard'
```
Comment (azumakuniyuki) [inline lib/sisimai/smtp/error.rb]: __[rubocop]__
* Layout/IndentationWidth: Use 2 (not -11) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

File: lib/sisimai/smtp/error.rb [hunk 3]
```diff
@@ -89,11 +89,13 @@ def soft_or_hard(argv1 = '', argv2 = '')
...
+              else
```
Comment (azumakuniyuki) [inline lib/sisimai/smtp/error.rb]: __[rubocop]__
* Layout/ElseAlignment: Align `else` with `if`.

File: spec/sisimai/smtp/error_spec.rb [hunk 4]
```diff
@@ -13,6 +13,10 @@
   isntbounces = ['delivered', 'feedback', 'vacation']
   dependondsn = ['undefined', 'onhold']
 
+  dependondsn_hard = [
```
Comment (azumakuniyuki) [inline spec/sisimai/smtp/error_spec.rb]: __[rubocop]__
* Lint/UselessAssignment: Useless assignment to variable - `dependondsn_hard`. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)

**Comment by azumakuniyuki** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31641851/issues/pull_request/84/diff/3d6b29b3619758c9e1462e89040ac24240a63e14?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: lib/sisimai/smtp/error.rb [hunk 5]
```diff
@@ -89,11 +89,13 @@ def soft_or_hard(argv1 = '', argv2 = '')
...
 
             if getchecked.nil?
-              # The value is not defined (returned undef)
               softorhard = ''
             else
-              # The value is 0 or 1
-              softorhard = getchecked == 1 ? 'hard' : 'soft'
+              softorhard if getchecked
+                'hard'
+              else
```
Comment (azumakuniyuki) [inline lib/sisimai/smtp/error.rb]: __[rubocop]__
* unexpected token kELSE
(Using Ruby 2.1 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)

**Comment by azumakuniyuki** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31641851/issues/pull_request/84/diff/9d820f8044f412caaf4e3e1989e6e6d72eeeb2fa?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: lib/sisimai/smtp/error.rb [hunk 6]
```diff
@@ -89,11 +89,13 @@ def soft_or_hard(argv1 = '', argv2 = '')
...
             if getchecked.nil?
-              # The value is not defined (returned undef)
               softorhard = ''
             else
-              # The value is 0 or 1
-              softorhard = getchecked == 1 ? 'hard' : 'soft'
+              softorhard = if getchecked
+                'hard'
+              else
+                'soft'
```
Comment (azumakuniyuki) [inline lib/sisimai/smtp/error.rb]: __[rubocop]__
* Layout/IndentationWidth: Use 2 (not -11) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

**Comment by azumakuniyuki** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31641851/issues/pull_request/84/diff/fb5a79680d40780672e1330a49874ca8ca5e957a?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

**Comment by azumakuniyuki** [comment]
Thanks for the pull-request. the correct value of `softbounce` on the parsed results for [set-of-emails/maildir/bsd/rfc3464-19.eml](https://github.com/sisimai/set-of-emails/blob/master/maildir/bsd/rfc3464-19.eml) is **1** because the value of `reason` is `onhold`.

**Comment by lunatyq** [comment]
@azumakuniyuki `Sisimai::SMTP::Error.is_permanent("5.0.901")` returns true

**Comment by azumakuniyuki** [comment]
@lunatyq The value of `softbounce` is decided on the definition in https://github.com/sisimai/rb-Sisimai/blob/master/lib/sisimai/smtp/error.rb#L11-L19 
```ruby
        SoftOrHard = {
          :soft => %w|
            blocked contenterror exceedlimit expired filtered mailboxfull
            mailererror mesgtoobig networkerror norelaying policyviolation
            rejected securityerror spamdetected suspend syntaxerror systemerror
            systemfull toomanyconn virusdetected
          |,
          :hard => %w|hasmoved hostunknown userunknown|
        }
```

Hard bounce is an error that an email will never been delivered to the destination host or recipient.

**Comment by lunatyq** [comment]
@azumakuniyuki but we have onhold reason and this check doesn't apply here. We check it earlier:
https://github.com/sisimai/rb-Sisimai/blob/master/lib/sisimai/smtp/error.rb#L87-L97

**Comment by azumakuniyuki** [comment]
@lunatyq Sorry, I completely forgot that `SoftOrHard` don't include `onhold`, `undefined` and other non-bounce reasons. Please disregard what I said at https://github.com/sisimai/rb-Sisimai/pull/84#issuecomment-325955934

**Comment by azumakuniyuki** [comment]
@lunatyq Would you like to include the following patch into this Pull-Request ?
```diff
--- spec/sisimai/bite/email/public-rfc3464_spec.rb	2017-08-20 17:25:39.000000000 +0900
+++ spec/sisimai/bite/email/public-rfc3464_spec.rb	2017-08-30 20:48:19.000000000 +0900
@@ -20,7 +20,7 @@
   { 'n' => '16', 's' => /\A5[.]0[.]\d+\z/,   'r' => /filtered/,    'a' => /RFC3464/, 'b' => /\A1\z/ },
   { 'n' => '17', 's' => /\A5[.]0[.]\d+\z/,   'r' => /expired/,     'a' => /RFC3464/, 'b' => /\A1\z/ },
   { 'n' => '18', 's' => /\A5[.]1[.]1\z/,     'r' => /userunknown/, 'a' => /RFC3464/, 'b' => /\A0\z/ },
-  { 'n' => '19', 's' => /\A5[.]0[.]\d+\z/,   'r' => /onhold/,      'a' => /RFC3464/, 'b' => /\A1\z/ },
+  { 'n' => '19', 's' => /\A5[.]0[.]\d+\z/,   'r' => /onhold/,      'a' => /RFC3464/, 'b' => /\A0\z/ },
   { 'n' => '20', 's' => /\A5[.]0[.]\d+\z/,   'r' => /mailererror/, 'a' => /RFC3464/, 'b' => /\A1\z/ },
   { 'n' => '21', 's' => /\A5[.]0[.]\d+\z/,   'r' => /networkerror/,'a' => /RFC3464/, 'b' => /\A1\z/ },
   { 'n' => '22', 's' => /\A5[.]0[.]\d+\z/,   'r' => /hostunknown/, 'a' => /RFC3464/, 'b' => /\A0\z/ },
```

**Comment by azumakuniyuki** [comment]
Thanks for the commit 26228ca, I'll merge the pull-request after the build test on Travis-CI has been succeeded.

**Pull request merged**